feat: Add primary key validation to dal:validate command

BrocksiNet opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us process your pull request.

Please make sure to fulfil our contribution guidelines (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be documented?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?

Entity definitions must have `PrimaryKey` flags that exactly match the database PRIMARY KEY constraint. When there's a mismatch, entity hydration fails silently, causing entities to not load properly. Additionally, the `DefinitionValidator` now validates this consistency to catch such issues early.

Fixes #13429

### 2. What does this change do, exactly?

This PR applies the primary key validation changes from PR #13430:

1. **Adds primary key validation to `DefinitionValidator`**: Validates that `PrimaryKey` flags in entity definitions match the actual `PRIMARY KEY` constraint in the database schema.

2. **Fixes primary key flag mismatches** in several entity definitions:
   - `ProductCrossSellingAssignedProductsDefinition`: Removed `PrimaryKey` flag from `ReferenceVersionField`
   - `ProductKeywordDictionaryDefinition`: Added `PrimaryKey` flag to `language_id` FkField
   - `ProductOptionDefinition`: Added `PrimaryKey` flag to `ReferenceVersionField`
   - `ProductPropertyDefinition`: Added `PrimaryKey` flag to `ReferenceVersionField`
   - `ProductSearchKeywordDefinition`: Added `PrimaryKey` flag to `language_id` FkField
   - `ProductStreamMappingDefinition`: Added `PrimaryKey` flag to `ReferenceVersionField`
   - `NumberRangeStateDefinition`: Added `PrimaryKey` flag to `IdField` and removed it from `FkField`

3. **Adds test coverage**: Created `DefinitionValidatorTest` and `DefinitionValidatorTestDefinition` to test the primary key consistency validation.

4. **Updates documentation**: Added information about the primary key validation feature to `RELEASE_INFO-6.7.md`.

### 3. Describe each step to reproduce the issue or behaviour.

Before this change:
1. Create an entity definition with `PrimaryKey` flags that don't match the database PRIMARY KEY
2. Run `bin/console dal:validate`
3. The mismatch would not be detected
4. Entity hydration would fail silently when trying to load entities

After this change:
1. Run `bin/console dal:validate`
2. The validator will detect and report any primary key mismatches
3. The error message clearly indicates which columns are mismatched

### 4. Please link to the relevant issues (if any).

- relates #13429
- Cherry-picked from #13430

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

BrocksiNet committed:
feat: Add primary key validation to dal:validate command

BrocksiNet committed:
feat: Add created_at and updated_at fields to DefinitionValidatorTestDefinition

BrocksiNet committed:
feat: Add test for primary key validation skipping non-StorageAware fields

BrocksiNet committed:
fix: Add ApiAware flag to language_id field in ProductKeywordDictionaryDefinition

BrocksiNet committed:
feat: Rename test definition

BrocksiNet committed:
chore: Remove DefinitionValidatorTestDefinitionWithNonStorageAwarePrimaryKey test class

nfortier-shopware commented on tests/unit/Core/Framework/DataAbstractionLayer/DefinitionValidatorTest.php:
```diff
@@ -0,0 +1,209 @@
+    {
+        yield 'matching primary key' => [
+            ['id'],
+        ];
```
Perhaps we do not need a provider for this?

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet committed:
refactor: simplify DefinitionValidatorTest by removing unused data provider and hardcoding primary keys

BrocksiNet committed:
refactor: remove deprecated PHPStan ignore comment from GoogleStorageFactory

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet commented on tests/unit/Core/Framework/DataAbstractionLayer/DefinitionValidatorTest.php:
```diff
@@ -0,0 +1,195 @@
+        $schemaManager = $this->createMock(AbstractSchemaManager::class);
+        $connection->method('createSchemaManager')->willReturn($schemaManager);
+        $schemaManager->method('introspectTable')->willThrowException(new \Exception('Table does not exist'));
+
```
👎🏼

mitelg reviewed (commented):
is this ready for review @BrocksiNet ?

BrocksiNet commented:
> is this ready for review @BrocksiNet ?

Yes, but I forgot about that PR, to be honest 🙈. And we need to update the release MD now... but feel free to provide feedback so we can finish it.

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -113,6 +113,11 @@ If the token has already been consumed, the user will be redirected directly to
 To support this behavior, a new `consumed` flag has been added to the payment token struct, which indicates if the token has already been processed.
 Payment tokens are no longer deleted immediately after use. A new scheduled task automatically removes expired tokens to keep the `payment_token` table clean.
 
+### Primary key validation in `dal:validate` command
```
needs to be moved to right version section (just as reminder)

mitelg commented on src/Core/Framework/DataAbstractionLayer/DefinitionValidator.php:
```diff
@@ -956,6 +958,57 @@ private function validateColumn(EntityDefinition $definition): array
+     *
+     * @return array<class-string<EntityDefinition>, list<string>>
+     */
+    private function validatePrimaryKeyConsistency(EntityDefinition $definition): array
```
would be nice, if we can wait until I finished the DBAL update, because this will also have effects on this method :grimacing:

BrocksiNet replied:
Fine for me. Just tell me when you're ready. Thanks.

BrocksiNet replied:
Just for reference: https://github.com/shopware/shopware/pull/14277

mitelg commented on src/Core/Framework/DataAbstractionLayer/Validation/TestDefinition/DefinitionValidatorTestDefinition.php:
```diff
@@ -0,0 +1,38 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\DataAbstractionLayer\Validation\TestDefinition;
```
why do we need to put it in this namespace?

BrocksiNet replied:
The test definitions must be in a non-test namespace because DefinitionValidator::validate() explicitly filters out definitions from test namespaces.

mitelg commented on tests/unit/Core/Framework/DataAbstractionLayer/DefinitionValidatorTest.php:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * @return \Generator<string, array{list<string>, list<string>}>
+     */
+    public static function primaryKeyMismatchProvider(): \Generator
```
can you move the provider next to the test method?

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet committed:
refactor: improvements after review

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

nfortier-shopware commented on src/Core/Framework/DataAbstractionLayer/DefinitionValidator.php:
```
            if ($this->shouldSkipDefinition($definitionClass)) {
                continue;
            }
```

and with
```
    /**
     * Check if a definition should be skipped during validation
     * @see https://regex101.com/r/hpxAVN/1
     */
    protected function shouldSkipDefinition(string $definitionClass): bool
    {
        return preg_match('/.*\\\\Tests?\\\\.*/', $definitionClass) || preg_match('/.*ComposerChild\\\\.*/', $definitionClass);
    }
````

nfortier-shopware replied:
Then you can isolate it in your Unit test, without this bad trick of storing fixtures/stubs in src

nfortier-shopware commented on src/Core/Framework/DataAbstractionLayer/Validation/TestDefinition/DefinitionValidatorTestDefinition.php:
```diff
@@ -0,0 +1,38 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\DataAbstractionLayer\Validation\TestDefinition;
```
This should belong to `namespace Shopware\Tests\Unit\Core\Framework\DataAbstractionLayer\Validation\Fixtures;`

(to resolve the problem of being blocked by the validate check see other comments)

nfortier-shopware commented on src/Core/Framework/DataAbstractionLayer/Validation/TestDefinition/DefinitionValidatorTestDefinition.php:
```diff
@@ -0,0 +1,38 @@
+ * Note: This is in a non-Test namespace so DefinitionValidator doesn't filter it out
+ */
+#[Package('framework')]
+class DefinitionValidatorTestDefinition extends EntityDefinition
```
It should be renamed as `DefinitionStub` instead

nfortier-shopware commented on src/Core/Framework/DataAbstractionLayer/Validation/TestDefinition/DefinitionValidatorWithNonStorageAwarePrimaryKeyTestDefinition.php:
```diff
@@ -0,0 +1,30 @@
+ * Used to test that the validator skips non-StorageAware fields when checking primary keys
+ */
+#[Package('framework')]
+class DefinitionValidatorWithNonStorageAwarePrimaryKeyTestDefinition extends DefinitionValidatorTestDefinition
```
`DefinitionWithNonStorageAwarePrimaryKeyStub` would give more clarity

nfortier-shopware commented on tests/unit/Core/Framework/DataAbstractionLayer/DefinitionValidatorTest.php:
```diff
@@ -0,0 +1,196 @@
+        $registry->method('getDefinitions')->willReturn([$definition]);
+        $registry->method('getByEntityName')->willReturn($definition);
+
+        return new DefinitionValidator($registry, $connection);
```
```
        // Create a custom validator that doesn't skip test definitions
        return new class($registry, $connection) extends DefinitionValidator {
            protected function shouldSkipDefinition(string $definitionClass): bool
            {
                // Don't skip test definitions for unit testing
                return false;
            }
        };
```

nfortier-shopware commented on tests/unit/Core/Framework/DataAbstractionLayer/DefinitionValidatorTest.php:
```diff
@@ -0,0 +1,196 @@
+        $registry->method('getDefinitions')->willReturn([$definition]);
+        $registry->method('getByEntityName')->willReturn($definition);
+
+        return new DefinitionValidator($registry, $connection);
```
same here

nfortier-shopware reviewed (changes requested):
We should not put fixtures in `src` when we can. 

Yes there are still some in src/Core/Test and Core/Framework/Test but it is also considered as sub test realm. We should not add more, and especially not in `Shopware\Core\Framework\DataAbstractionLayer\Validation\TestDefinition;`

Instead the class should be testable, since we are changing it in this PR, please see the suggestions to isolate the check about test namespace

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet committed:
refactor: simplify definition validation logic by extracting skip condition to a dedicated method and update tests to use new definition stubs

mitelg reviewed (commented):
looks good to me, only conflicts need to be resolved. 

hopefully I can finish the DBAL update first :grimacing:

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/shopware/shopware?pullRequest=13483) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/shopware/shopware?pullRequest=13483) before we can accept your contribution.<br/>**17** out of **18** committers have signed the CLA.<br/><br/>:white_check_mark: nfortier-shopware<br/>:white_check_mark: BrocksiNet<br/>:white_check_mark: keulinho<br/>:white_check_mark: mstegmeyer<br/>:white_check_mark: larskemper<br/>:white_check_mark: En0Ma1259<br/>:white_check_mark: h1k3r<br/>:white_check_mark: FlorianKe<br/>:white_check_mark: jleifeld<br/>:white_check_mark: davidtraum<br/>:white_check_mark: Alex--C<br/>:white_check_mark: SebastianFranze<br/>:white_check_mark: bschulzebaek<br/>:white_check_mark: aragon999<br/>:white_check_mark: pt1602<br/>:white_check_mark: timtheisinger<br/>:white_check_mark: ffrank913<br/>:x: shopware-octo-sts-app[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/shopware/shopware?pullRequest=13483) it.</sub>

BrocksiNet committed:
Merge remote-tracking branch 'origin/trunk' into feat/validation-primary-key

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

h1k3r commented on src/Core/Framework/DataAbstractionLayer/DefinitionValidator.php:
```diff
@@ -956,6 +968,57 @@ private function validateColumn(EntityDefinition $definition): array
+        // Skip if table doesn't exist yet
+        try {
+            $table = $this->connection->createSchemaManager()->introspectTable($definition->getEntityName());
+        } catch (\Exception) {
```
Does it make sense to narrow exception type, to avoid silently skipping check in case of some valid errors? 🤔

mitelg commented on src/Core/Framework/DataAbstractionLayer/DefinitionValidator.php:
```diff
@@ -956,6 +968,57 @@ private function validateColumn(EntityDefinition $definition): array
+
+        // Skip if table doesn't exist yet
+        try {
+            $table = $this->connection->createSchemaManager()->introspectTable($definition->getEntityName());
```
I refactored the validator, so please pass the `$schema` into this private method. it already contains the whole schema, so no need to introspect again here

BrocksiNet committed:
Merge branch 'trunk' into feat/validation-primary-key

BrocksiNet committed:
feat: Enhance primary key validation in `dal:validate` command to prevent entity hydration failures

- Added validation to detect mismatches between database PRIMARY KEY constraints and entity definition PrimaryKey flags.
- Improved error messaging for mismatches to aid in debugging.
- Updated `DefinitionValidator` to utilize schema introspection for primary key checks.
- Adjusted unit tests to reflect changes in validation logic.
